File: dp/Longest Arithmetic Series.cpp
#include<bits/stdc++.h>
using namespace std;
int main(){
    string s;
    while(getline(cin,s)){
        stringstream ss;
        ss<<s;
        int k=0,arr[50],dp[50];
        while(ss>>arr[k++]);
        sort(arr,arr+k);
        for(int i=0;i<=k;i++)cout<<arr[i]<<'\n';
        for(int i=1;i<k;i++)cout<<i<<" "<< dp[i]<<"\n";
        int ma=1,cnt=0;
        for(int i=1;i<k-1;i++){
            if(dp[i]==dp[i+1])cnt++;
            else ma=max(ma,cnt),cnt=0;
        }
        cout<<ma+1<<'\n';
    }
}

File: dp/Longest Arithmetic Series.py
def solve(A):
        ans = 2       
        n= len(A)
        if n<=2 :
            return n
        llap = [2]*n
        A.sort()
   
        for j in range(n-2, -1, -1):
            i= j-1
            k= j+1
            while(i>=0 and k<n):
                if A[i]+A[k] == 2*A[j]:
                    llap[j] = max(llap[k]+1,llap[j])
                    ans = max(ans, llap[j])
                    i-=1
                    k+=1
                elif A[i]+A[k] < 2*A[j]:
                    k += 1
                else:
                    i -= 1
   
        return ans 
t = 5
while True:
    if t == 0:
        break
    t -= 1
    arr = [int(ele) for ele in input().split()]
    print(solve(arr))

